fix: improve password validation flow

defelmnq opened the pull request:
Refers to #14984 

Currently, password validation is done backend side and is not explicit enough so it can be painful to create first users.
We'd like to make this validation easier - but also duplicate it frontend side to make it smoother.

Flows involved : 
- First user set password
- New user set password
- Change password

defelmnq committed:
fix(setup): improve password validation flow on first user setup

defelmnq committed:
fix(setup): improve password validation flow on first user setup

johnstcn commented on coderd/userpassword/userpassword_test.go:
```diff
@@ -14,6 +15,32 @@ import (
 
 func TestUserPassword(t *testing.T) {
 	t.Parallel()
+
+	t.Run("Invalid - Too short password", func(t *testing.T) {
+		t.Parallel()
+		err := userpassword.Validate("pass")
+		require.Error(t, err)
+	})
+
+	t.Run("Invalid - Too long password", func(t *testing.T) {
+		t.Parallel()
+
+		var sb strings.Builder
+		for i := 0; i < 65; i++ {
+			sb.WriteString("a")
+		}
+
+		err := userpassword.Validate(sb.String())
+		require.Error(t, err)
+	})
+
+	t.Run("Ok", func(t *testing.T) {
+		t.Parallel()
+
+		err := userpassword.Validate("CorrectPassword")
+		require.NoError(t, err)
+	})
+
```
suggestion: this is a perfect candidate for a table-driven test

johnstcn reviewed (commented):
I'm in two minds about this change.

On the one hand, I agree that the entropy-based approach is obtuse and annoying, and that it is in some way the user's responsibility to have a minimum password security value. Also, if security is your top priority, password authentication is maybe not the best approach.

On the other hand, allowing folks to just use `password` or `abc123` feels wrong to me, and this seems like the kind of thing a security audit would find deeply wanting.

I considered the below rules:
1) Mandate use of at least one character from each of the following classes: `[a-z]` `[A-Z]` `[0-9]`
2) Mandate minimum length 9

However this would immediately discount passwords consisting entirely of non-English υηι¢σ∂є ¢нαяα¢тєяѕ into account.


Here's an idea:  given that the entropy calculation check is relatively cheap, would it be a terrible idea to expose an endpoint to validate this? The frontend could then consume this and show the entropy calculation transparently.

EDIT: alternative proposition: adjust the rule-based approach to

1) At least one non-alphanumeric non-whitespace character (e.g. `[^a-zA-Z0-9\s]+`)
2) Minimum length of 9

defelmnq committed:
feat(password): WIP

defelmnq committed:
feat(password): apply backend logic to all password set fields

BrunoQuaresma commented:
I think the way you did it is fine, but I’d suggest trying to use Yup features to simplify it since we’re already using it. Reposting from Slack:

> Ok, cool. The good thing is, you can already do something like that using [Yup](https://github.com/jquense/yup), the validation library we use. You can check out an example on [this StackOverflow question](https://stackoverflow.com/questions/55811114/async-validation-with-formik-yup-and-react). You can probably place it in the formUtils.ts file.

Sorry for the late message 🙏

defelmnq commented:
> I think the way you did it is fine, but I’d suggest trying to use Yup features to simplify it since we’re already using it. Reposting from Slack:
> 
> > Ok, cool. The good thing is, you can already do something like that using [Yup](https://github.com/jquense/yup), the validation library we use. You can check out an example on [this StackOverflow question](https://stackoverflow.com/questions/55811114/async-validation-with-formik-yup-and-react). You can probably place it in the formUtils.ts file.
> 
> Sorry for the late message 🙏

Hey @BrunoQuaresma thanks for the review and validation on this piece of code. A bit more context [can be found here](https://github.com/coder/coder/issues/14984) but globally the idea is to keep frontend validation as small as possible - (probably just for the obvious things) and then have a debounced backend one for the more detailed validation - so we ensure consistent validation everywhere.

I'll put the same kind of changes in all other password fields - so the PR will be fully reviewable.

defelmnq commented on coderd/userpassword/userpassword.go:
```diff
@@ -138,10 +137,8 @@ func hashWithSaltAndIter(password string, salt []byte, iter int) string {
-	err := passwordvalidator.Validate(password, 52)
-	if err != nil {
-		return err
+	if len(password) < 6 {
```
For now this logic stays pretty simple - here's the place where we'll be able to apply all the validation logic we want

johnstcn replied:
As far as I'm concerned here, we are squishing together three distinct separate changes:
1) Expose a password complexity validation endpoint on the backend
2) Modify the password validation logic on the FE to query the BE
3) **Modify the password validation logic on the BE**

Of these three options, 1) and 2) are relatively innocuous. 3) is the one I'm mainly concerned about.

I propose keeping the existing password validation logic (via entropy) in this PR and opening a separate PR to modify the password validation logic.

defelmnq committed:
Merge remote-tracking branch 'origin/main' into fix-password-validation

defelmnq committed:
Merge remote-tracking branch 'origin/main' into fix-password-validation

defelmnq committed:
feat(password): apply backend logic to all password set fields

defelmnq committed:
feat(password): apply backend logic to all password set fields

defelmnq committed:
feat(password): apply backend logic to all password set fields

defelmnq committed:
feat(password): apply backend logic to all password set fields

defelmnq committed:
feat(password): apply backend logic to all password set fields

defelmnq committed:
feat(password): add test for validate password method

defelmnq commented:
PR is for me ready for a first review - both frontend and backend side.

We'd just need to : 
- Define what we define as a strong enough password security wise cc @johnstcn (?) and change the rules based on that.
- Potentially improve the wording frontend side.

defelmnq committed:
fix(password): display only if password is set

johnstcn commented on coderd/userpassword/userpassword_test.go:
```diff
@@ -5,53 +5,73 @@
+		password string
+		wantErr  bool
+	}{
+		{"Invalid - Too short password", "pass", true},
+		{"Invalid - Too long password", strings.Repeat("a", 65), true},
+		{"Ok", "CorrectPassword", false},
```
nit: I have a weak preference for test cases with keyed struct fields

johnstcn commented on coderd/users_test.go:
```diff
@@ -1084,20 +1084,22 @@ func TestUpdateUserPassword(t *testing.T) {
-		})
-		require.Error(t, err, "member should not be able to update own password without providing old password")
-	})
+	// FIXME:  Re-enable the tests once real logic changed
```
instead of commenting out a test, add a `t.Skip()`

defelmnq committed:
WIP: Working on testing improvement

defelmnq committed:
site: change logic to generate error instead of helper text on password validation

defelmnq committed:
fix storybook

defelmnq commented:
Based on conversation @johnstcn - I changed a bit the PR. 

We pushed back the validation of the password as a prerequisite. If we then want it to just be a hint, it will be done in another PR.

Also we continue to use entropy-based validation as before.

johnstcn commented on coderd/coderd.go:
```diff
@@ -991,6 +991,7 @@ func New(options *Options) *API {
 				r.Use(httpmw.RateLimit(options.LoginRateLimit, time.Minute))
 				r.Post("/login", api.postLogin)
 				r.Post("/otp/request", api.postRequestOneTimePasscode)
+				r.Post("/validate-password", api.validateUserPassword)
```
suggestion: add a rate limit here, could also be a follow-up until we know a reasonable value

defelmnq replied:
Yeah good one - let me add a follow-up for this one 🙏

johnstcn commented on coderd/userauth.go:
```diff
@@ -437,6 +437,38 @@ func (api *API) postChangePasswordWithOneTimePasscode(rw http.ResponseWriter, r
+
+	err := userpassword.Validate(req.Password)
+	if err != nil {
+		valid = false
```
suggestion: We could return the error in the response for the user to adjust their input. However, we should make sure that it does not contain the original password from the request.

johnstcn commented on coderd/userpassword/userpassword_test.go:
praise: nice tests

johnstcn commented on coderd/users_test.go:
```diff
@@ -1084,7 +1084,11 @@ func TestUpdateUserPassword(t *testing.T) {
+	// FIXME: Re-enable the tests once real logic changed
+	// Currently there's no check in code to validate that users have to put the old password
 	t.Run("MemberCantUpdateOwnPasswordWithoutOldPassword", func(t *testing.T) {
+		t.Skip()
```
Isn't this fixed now?

johnstcn commented on codersdk/users.go:
```diff
@@ -176,6 +176,14 @@ type UpdateUserProfileRequest struct {
+}
+
+type ValidateUserPasswordResponse struct {
+	Valid bool `json:"valid"`
```
suggestion: as above, if we include the error it will be easier for the user to adjust their input.

BrunoQuaresma commented:
> I am not sure yet how validation should be done for storybook ?

Here are the steps that I would take:

- Create an "invalid password" story for each component that uses it
- Add an interaction test
- Mock the API call to return the API error
- Expect to see if the message is correctly displayed

You can see an example of a  "test like this"[here](https://github.com/coder/coder/blob/d89ecebb4e5e7260ec1e8bee20cae503fd0adc66/site/src/modules/workspaces/WorkspaceTiming/WorkspaceTimings.stories.tsx). Please, let me know if you this is helpful enough or if you have any additional questions 🙏 .

defelmnq committed:
Merge remote-tracking branch 'origin/main' into fix-password-validation

defelmnq committed:
feat(password): add details field to validate password endpoint

defelmnq committed:
feat(password): add details field to validate password endpoint

defelmnq commented:
Just sharing some screenshot of what it currently looks like for visibility.

![Screenshot 2024-10-28 at 16 55 13](https://github.com/user-attachments/assets/f0bed424-1da0-4dc1-94f6-251f9100bed7)
![Screenshot 2024-10-28 at 16 54 30](https://github.com/user-attachments/assets/cc267e5d-e9ba-4612-ac91-1b361e6db7ef)
![Screenshot 2024-10-28 at 16 55 04](https://github.com/user-attachments/assets/94c80131-7bf1-4c47-92af-269806c8a1ce)

defelmnq commented:
@johnstcn @BrunoQuaresma 

- I added some frontend tests, not sure how to do for all of them but tried to add some. 
- Overall, I also added the details logic described by Cian - I've not added any logic to hide the user's password yet, because the library itself describe the error as being safe to display for user, but I can add it if you think it's better.

johnstcn commented on site/src/pages/UserSettingsPage/SecurityPage/SecurityPage.test.tsx:
```diff
@@ -76,24 +76,6 @@ test("update password with incorrect old password", async () => {
 	expect(API.updateUserPassword).toBeCalledWith(user.id, newSecurityFormValues);
 });
 
-test("update password with invalid password", async () => {
-	jest.spyOn(API, "updateUserPassword").mockRejectedValueOnce(
-		mockApiError({
-			message: "Invalid password.",
-			validations: [{ detail: "Invalid password.", field: "password" }],
-		}),
-	);
-
-	const { user } = await renderPage();
-	fillAndSubmitSecurityForm();
-
-	const errorMessage = await screen.findAllByText("Invalid password.");
-	expect(errorMessage).toBeDefined();
-	expect(errorMessage).toHaveLength(2);
-	expect(API.updateUserPassword).toBeCalledTimes(1);
-	expect(API.updateUserPassword).toBeCalledWith(user.id, newSecurityFormValues);
-});
-
```
I think we instead modify this test to spy on the validate user password endpoint.

johnstcn reviewed (approved):
Backed changes look good to me! @BrunoQuaresma could you take a look at the FE changes?

Kira-Pilot commented on site/src/api/api.ts:
```diff
@@ -1322,6 +1322,15 @@ class ApiMethods {
 	};
 
+	validateUserPassword = async (
+		password: string,
```
should be using the `ValidateUserPasswordRequest` interface here?

Kira-Pilot commented on site/src/pages/CreateUserPage/CreateUserForm.tsx:
```diff
@@ -63,6 +63,8 @@ export const authMethodLanguage = {
 export interface CreateUserFormProps {
 	onSubmit: (user: TypesGen.CreateUserRequestWithOrgs) => void;
 	onCancel: () => void;
+	onPasswordChange: (password: string) => void;
```
I think we can also use `ValidateUserPasswordRequest` here

Kira-Pilot commented on site/src/api/queries/users.ts:
```diff
@@ -9,6 +9,7 @@ import type {
 	UpdateUserProfileRequest,
 	User,
 	UsersRequest,
+	ValidateUserPasswordRequest,
```
Where are we using this?

Kira-Pilot commented on site/src/pages/CreateUserPage/CreateUserForm.tsx:
```diff
@@ -104,6 +114,10 @@ export const CreateUserForm: FC<
 
+	useEffect(() => {
+		onPasswordChange?.(form.values.password);
+	}, [form.values.password, onPasswordChange]); // Run effect when password changes
```
I think we want to avoid `useEffect` as much as possible - it causes unexpected rerenders, which, decoupled with your debouncing function, are going to be difficult to troubleshoot. Instead, I suggest you add an `onChange` attribute to the `TextField` component used for the password input. Something like:
```
onChange={(e) => {
    form.handleChange(e);
    handlePasswordChange(e.target.value);
}}
```

If you search in the file you can see us using this pattern for other inputs.

Kira-Pilot commented on site/src/pages/CreateUserPage/CreateUserPage.tsx:
```diff
@@ -17,6 +18,25 @@ export const CreateUserPage: FC = () => {
 	const authMethodsQuery = useQuery(authMethods());
+	const validatePasswordMutation = useMutation(validatePassword());
+
+	const [passwordValidator, setPasswordValidator] = useState({
```
Is there a reason all this logic must live in the parent component instead of directly inside `CreateUserForm`? I could be missing something and @BrunoQuaresma might be able to weigh in.

But if this logic can live in the form component itself, I think it should, for simplicity.

Kira-Pilot replied:
Scanning down through the PR, I see that we're duplicating this logic across several views. To avoid doing this, I think we should create a hook called `useValidatePassword` or something similar, and then call that hook across these components.

I suggest pairing with @BrunoQuaresma or @aslilac if you need help here!

Kira-Pilot reviewed (dismissed):
Thanks for taking a look at this user improvement and nice job committing some full stack code!

I left some comments on your FE changes. Might be easiest to hop in Discord with one of our UI engineers. LMK if you have any questions!

aslilac commented on coderd/userauth.go:
```diff
@@ -437,6 +437,41 @@ func (api *API) postChangePasswordWithOneTimePasscode(rw http.ResponseWriter, r
+// @Param request body codersdk.ValidateUserPasswordRequest true "Validate user password request"
+// @Success 200 {object} codersdk.ValidateUserPasswordResponse
+// @Router /users/validate-password [post]
+func (*API) validateUserPassword(rw http.ResponseWriter, r *http.Request) {
```
nit about referring to all of this stuff as "user password" validation. the password is not _yet_ attached to a user, so it's not a user password, it's just a password, so I'd expect it to be called `validatePassword` and so on.

BrunoQuaresma committed:
Extract validation logic to a component

BrunoQuaresma commented:
@defelmnq  what do you think about the changes? I would also appreciate a QA from you just to ensure I didn't miss any use cases.

defelmnq committed:
Merge remote-tracking branch 'origin/main' into fix-password-validation

BrunoQuaresma reviewed (approved):
I’m self-approving this for now to avoid blocking the PR any longer. If needed, I can take any follow-up actions later on. @defelmnq it is up to you when to merge this PR 🙏

EdwardAngert commented:
#15382 fixes the errors in the `weekly-docs` check that's currently failing
